libr/anal/type.c

```diff
@@ -122,18 +122,238 @@ R_API void r_anal_save_parsed_type(RAnal *anal, const char *parsed) {
 static int typecmp(const void *a, const void *b) {
 	char *type1 = (char *) a;
 	char *type2 = (char *) b;
-	return strcmp (type1 , type2);
+	return strcmp (type1, type2);
 }
 
 R_API RList *r_anal_types_from_fcn(RAnal *anal, RAnalFunction *fcn) {
 	RListIter *iter;
 	RAnalVar *var;
 	RList *list = r_anal_var_all_list (anal, fcn);
-        RList *type_used = r_list_new ();
-        r_list_foreach (list, iter, var) {
-		r_list_append (type_used , var->type);
-        }
-	RList *uniq = r_list_uniq (type_used , typecmp);
+	RList *type_used = r_list_new ();
+	r_list_foreach (list, iter, var) {
+		r_list_append (type_used, var->type);
+	}
+	RList *uniq = r_list_uniq (type_used, typecmp);
 	r_list_free (type_used);
 	return uniq;
 }
+
+static void enum_type_fini(void *e, void *user) {
+	(void)user;
+	RAnalEnumCase *cas = e;
+	free ((char *)cas->name);
+}
+
+static RAnalBaseType *get_enum_type(RAnal *anal, const char *sanitized_name) {
+	r_return_val_if_fail (sanitized_name && anal, NULL);
+
+	RAnalBaseType *base_type = R_NEW0 (RAnalBaseType);
+	if (!base_type) {
+		return NULL;
+	}
+	base_type->kind = R_ANAL_BASE_TYPE_KIND_ENUM;
+	
+	RAnalBaseTypeEnum base_enum;
+	char *key = r_str_newf ("enum.%s", sanitized_name);
+	if (!key) {
+		free (base_type);
+		return NULL;
+	}
+	const char *members = sdb_const_get (anal->sdb_types, key, NULL);
+	free (key);
+
+	RVector cases;
+	r_vector_init (&cases, sizeof (RAnalEnumCase), enum_type_fini, NULL);
+
+	if (!r_vector_reserve (&cases, (size_t) sdb_alen (members))) {
+		free (base_type);
+		r_vector_free (&cases);
+		return NULL;
+	}
+
+	char *cur;
+	sdb_aforeach (cur, members) {
+		char *val_key = r_str_newf ("enum.%s.%s", sanitized_name, cur);
+		if (!val_key) {
+			free (base_type);
```

Comment:
return free_locals(base_type, &cases); or use goto to avoid repeating the same frees before each return statement